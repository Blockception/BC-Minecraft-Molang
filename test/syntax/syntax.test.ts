import { Types } from "bc-minecraft-bedrock-types";
import { parseMolang } from "../../src/molang/syntax/parse";
import { valid_syntaxes } from "../data/dataset-valid";
import { ExpressionNode, NodeType } from "../../src/molang/syntax/nodes";
import { invalid_syntaxes } from "../data/dataset-invalid";
import { MolangSyntaxError } from "../../src/molang";

describe("molang - syntax", () => {
  describe("should be able to parse and match the syntax tree generated", () => {
    test.each(valid_syntaxes)("%#. %s", (s) => {
      const n = parseMolang(Types.OffsetWord.create(s, 0));
      n.forEach(cleanupNodes);
      expect(n).toMatchSnapshot();

      n.forEach(validateNode);
    });
  });

  describe("should throw an error", () => {
    test.each(invalid_syntaxes)("%#. %s", (s) => {
      expect(() => parseMolang(Types.OffsetWord.create(s, 0))).toThrow(MolangSyntaxError);
    });
  });
});

function cleanupNodes(node: ExpressionNode) {
  switch (node.type) {
    case NodeType.StatementSequence:
      node.statements.forEach(cleanupNodes);
      break;
    case NodeType.ArrayAccess:
      [node.array, node.index].forEach(cleanupNodes);
      break;
    case NodeType.Assignment:
    case NodeType.BinaryOperation:
      [node.left, node.right].forEach(cleanupNodes);
      break;
    case NodeType.FunctionCall:
      node.arguments.forEach(cleanupNodes);
      break;
    case NodeType.UnaryOperation:
      cleanupNodes(node.operand);
      break;
    case NodeType.Conditional:
      [node.condition, node.falseExpression, node.trueExpression]
        .filter((item) => item !== undefined)
        .forEach(cleanupNodes);
      break;
  }

  // Convert type to string for each identificatio
  (node as any).type = NodeType[node.type];
}

function validateNode(node: ExpressionNode): void {
  expect(node).toBeDefined();
  expect(node).toHaveProperty("type");
  expect(node).toHaveProperty("position");

  switch (node.type) {
    case NodeType.StatementSequence:
      node.statements.forEach(validateNode);
      throw new Error(`found a statement object: ${JSON.stringify(node)}`);
      break;
    case NodeType.ArrayAccess:
      [node.array, node.index].forEach(validateNode);
      break;
    case NodeType.Assignment:
    case NodeType.BinaryOperation:
      [node.left, node.right].forEach(validateNode);
      break;
    case NodeType.FunctionCall:
      node.arguments.forEach(validateNode);
      break;
    case NodeType.UnaryOperation:
      validateNode(node.operand);
      break;
    case NodeType.Conditional:
      [node.condition, node.falseExpression, node.trueExpression]
        .filter((item) => item !== undefined)
        .forEach(validateNode);
      break;
    case NodeType.Marker:
      throw new Error(`found a marker object: ${JSON.stringify(node)}`);
    case NodeType.Literal:
    case NodeType.NullishCoalescing:
    case NodeType.ResourceReference:
    case NodeType.StringLiteral:
    case NodeType.Variable:
      break;
  }
}
