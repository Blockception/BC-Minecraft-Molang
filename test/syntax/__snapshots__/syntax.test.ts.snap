// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`molang - syntax brackets are hard 0. 1.5 - math.pow(1.5, 1.0 - 015.0 * query.modified_move_speed) * query.camera_rotation(0) 1`] = `
[
  "surround: ",
  "1.5 [number]",
  "- [operator]",
  "math [text]",
  ". [punction]",
  "pow [text]",
  [
    "surround: ()",
    "1.5 [number]",
    ", [punction]",
    "1.0 [number]",
    "- [operator]",
    "015.0 [number]",
    "* [operator]",
    "query [text]",
    ". [punction]",
    "modified_move_speed [unknown]",
  ],
  ") [unknown]",
  "* [operator]",
  "query [text]",
  ". [punction]",
  "camera_rotation [unknown]",
  [
    "surround: ()",
    "0 [number]",
  ],
  [
    "surround: ",
    ") [unknown]",
  ],
]
`;

exports[`molang - syntax brackets are hard 1. math.clamp(math.sqrt(math.pow(query.position_delta(0), 2.0) + math.pow(query.position_delta(2), 2.0)), 0.0, 0.1) 1`] = `
[
  "surround: ",
  "math [keyword]",
  ". [access]",
  "clamp [identifier]",
  [
    "surround: ()",
    "math [keyword]",
    ". [access]",
    "sqrt [identifier]",
    [
      "surround: ()",
      "math [keyword]",
      ". [access]",
      "pow [identifier]",
      [
        "surround: ()",
        "query [keyword]",
        ". [access]",
        "position_delta [identifier]",
        [
          "surround: ()",
          "0 [number]",
        ],
        [
          "surround: ",
          ") [unknown]",
          ", [punction]",
          "2.0 [number]",
        ],
      ],
      ") [unknown]",
      "+ [operator]",
      "math [text]",
      ". [punction]",
      "pow [text]",
      [
        "surround: ()",
        "query [keyword]",
        ". [access]",
        "position_delta [identifier]",
        [
          "surround: ()",
          "2 [number]",
        ],
        [
          "surround: ",
          ") [unknown]",
          ", [punction]",
          "2.0 [number]",
        ],
      ],
      [
        "surround: ",
        ") [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      ", [punction]",
      "0.0 [number]",
      ", [punction]",
      "0.1 [number]",
    ],
  ],
  [
    "surround: ",
    ") [unknown]",
  ],
]
`;

exports[`molang - syntax brackets are hard 2. -175 * math.sin(variable.attack_time * 180) 1`] = `
[
  "surround: ",
  "- [operator]",
  "175 [number]",
  "* [operator]",
  "math [text]",
  ". [punction]",
  "sin [text]",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "attack_time [identifier]",
    "* [operator]",
    "180 [number]",
  ],
  [
    "surround: ",
    ") [unknown]",
  ],
]
`;

exports[`molang - syntax brackets are hard 3. array.skins[query.property('minecraft:has_nectar') + query.is_angry * 2] 1`] = `
[
  "surround: ",
  "array [keyword]",
  ". [access]",
  "skins [identifier]",
  [
    "surround: []",
    "query [keyword]",
    ". [access]",
    "property [identifier]",
    [
      "surround: ()",
      "'minecraft:has_nectar' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "+ [operator]",
      "query [text]",
      ". [punction]",
      "is_angry [unknown]",
      "* [operator]",
      "2 [number]",
    ],
  ],
  [
    "surround: ",
    "] [unknown]",
  ],
]
`;

exports[`molang - syntax brackets are hard 4. variable.x_head_rotation = math.clamp(query.target_x_rotation, -32.5, 32.5); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "x_head_rotation [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "math [text]",
    ". [punction]",
    "clamp [text]",
    [
      "surround: ()",
      "query [keyword]",
      ". [access]",
      "target_x_rotation [identifier]",
      ", [punction]",
      "- [operator]",
      "32.5 [number]",
      ", [punction]",
      "32.5 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 5. variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "y_head_rotation [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "math [text]",
    ". [punction]",
    "clamp [text]",
    [
      "surround: ()",
      "query [keyword]",
      ". [access]",
      "target_y_rotation [identifier]",
      ", [punction]",
      "- [operator]",
      "22.5 [number]",
      ", [punction]",
      "25 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 6. variable.is_rolled_up = query.property('minecraft:armadillo_state') != 'unrolled'; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "is_rolled_up [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "property [text]",
    [
      "surround: ()",
      "'minecraft:armadillo_state' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "!= [compare]",
      "'unrolled' [text]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 7. variable.rolled_up_time = variable.is_rolled_up ? ((variable.rolled_up_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "rolled_up_time [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "variable [text]",
    ". [punction]",
    "is_rolled_up [unknown]",
    "? [compare]",
    [
      "surround: ()",
      [
        "surround: ()",
        "variable [keyword]",
        ". [access]",
        "rolled_up_time [identifier]",
        "?? [operator]",
        "0.0 [number]",
      ],
      [
        "surround: ",
        ") [unknown]",
        "+ [operator]",
        "query [text]",
        ". [punction]",
        "delta_time [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      ": [operator]",
      "0.0 [number]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 8. variable.is_peeking = query.property('minecraft:armadillo_state') == 'rolled_up_peeking'; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "is_peeking [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "property [text]",
    [
      "surround: ()",
      "'minecraft:armadillo_state' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "== [compare]",
      "'rolled_up_peeking' [text]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 9. variable.is_unrolling = query.property('minecraft:armadillo_state') == 'rolled_up_unrolling'; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "is_unrolling [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "property [text]",
    [
      "surround: ()",
      "'minecraft:armadillo_state' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "== [compare]",
      "'rolled_up_unrolling' [text]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 10. variable.unrolling_time = variable.is_unrolling ? ((variable.unrolling_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "unrolling_time [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "variable [text]",
    ". [punction]",
    "is_unrolling [unknown]",
    "? [compare]",
    [
      "surround: ()",
      [
        "surround: ()",
        "variable [keyword]",
        ". [access]",
        "unrolling_time [identifier]",
        "?? [operator]",
        "0.0 [number]",
      ],
      [
        "surround: ",
        ") [unknown]",
        "+ [operator]",
        "query [text]",
        ". [punction]",
        "delta_time [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      ": [operator]",
      "0.0 [number]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 11. variable.use_rolled_up_model = variable.rolled_up_time >= 0.2083 && (variable.unrolling_time == 0.0 || variable.unrolling_time <= 1.25 ); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "use_rolled_up_model [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "variable [text]",
    ". [punction]",
    "rolled_up_time [unknown]",
    ">= [compare]",
    "0.2083 [number]",
    "&& [operator]",
    [
      "surround: ()",
      "variable [keyword]",
      ". [access]",
      "unrolling_time [identifier]",
      "== [compare]",
      "0.0 [number]",
      "|| [operator]",
      "variable [text]",
      ". [punction]",
      "unrolling_time [unknown]",
      "<= [compare]",
      "1.25 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 12. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "walk_anim_time_update [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "anim_time [unknown]",
    "+ [operator]",
    "math [text]",
    ". [punction]",
    "min [text]",
    [
      "surround: ()",
      "3.0 [number]",
      ", [punction]",
      "math [text]",
      ". [punction]",
      "lerp [text]",
      [
        "surround: ()",
        "2.0 [number]",
        ", [punction]",
        "5.0 [number]",
        ", [punction]",
        "query [text]",
        ". [punction]",
        "modified_move_speed [unknown]",
      ],
      [
        "surround: ",
        ") [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      "* [operator]",
      "query [text]",
      ". [punction]",
      "delta_time [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 13. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time;variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "walk_anim_time_update [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "anim_time [unknown]",
    "+ [operator]",
    "math [text]",
    ". [punction]",
    "min [text]",
    [
      "surround: ()",
      "3.0 [number]",
      ", [punction]",
      "math [text]",
      ". [punction]",
      "lerp [text]",
      [
        "surround: ()",
        "2.0 [number]",
        ", [punction]",
        "5.0 [number]",
        ", [punction]",
        "query [text]",
        ". [punction]",
        "modified_move_speed [unknown]",
      ],
      [
        "surround: ",
        ") [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      "* [operator]",
      "query [text]",
      ". [punction]",
      "delta_time [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax brackets are hard 13. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time;variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 2`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "y_head_rotation [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "math [text]",
    ". [punction]",
    "clamp [text]",
    [
      "surround: ()",
      "query [keyword]",
      ". [access]",
      "target_y_rotation [identifier]",
      ", [punction]",
      "- [operator]",
      "22.5 [number]",
      ", [punction]",
      "25 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 0. variable.agent.anim_state == 2 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "agent [identifier]",
    ". [access]",
    "anim_state [identifier]",
  ],
  "== [compare]",
  [
    "surround: ()",
    "2 [number]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 1. variable.agent.anim_state > 2 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "agent [identifier]",
    ". [access]",
    "anim_state [identifier]",
  ],
  "> [compare]",
  [
    "surround: ()",
    "2 [number]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 2. variable.agent.anim_state < 2 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "agent [identifier]",
    ". [access]",
    "anim_state [identifier]",
  ],
  "< [compare]",
  [
    "surround: ()",
    "2 [number]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 3. variable.agent.anim_state != 2 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "agent [identifier]",
    ". [access]",
    "anim_state [identifier]",
  ],
  "!= [compare]",
  [
    "surround: ()",
    "2 [number]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 4. variable.agent.anim_state >= 2 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "agent [identifier]",
    ". [access]",
    "anim_state [identifier]",
  ],
  ">= [compare]",
  [
    "surround: ()",
    "2 [number]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 5. variable.agent.anim_state <= 2 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "agent [identifier]",
    ". [access]",
    "anim_state [identifier]",
  ],
  "<= [compare]",
  [
    "surround: ()",
    "2 [number]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 6. !variable.swaying && query.all_animations_finished 1`] = `
[
  "surround: ",
  "! [operator]",
  "variable [keyword]",
  ". [access]",
  "swaying [identifier]",
  "&& [operator]",
  "query [text]",
  ". [punction]",
  "all_animations_finished [unknown]",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 7. 1.5 - math.pow(1.5, 1.0 - 015.0 * query.modified_move_speed) * query.camera_rotation(0) 1`] = `
[
  "surround: ",
  "1.5 [number]",
  "- [operator]",
  "math [text]",
  ". [punction]",
  "pow [text]",
  [
    "surround: ()",
    "1.5 [number]",
    ", [punction]",
    "1.0 [number]",
    "- [operator]",
    "015.0 [number]",
    "* [operator]",
    "query [text]",
    ". [punction]",
    "modified_move_speed [unknown]",
  ],
  ") [unknown]",
  "* [operator]",
  "query [text]",
  ". [punction]",
  "camera_rotation [unknown]",
  [
    "surround: ()",
    "0 [number]",
  ],
  [
    "surround: ",
    ") [unknown]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 8. math.clamp(math.sqrt(math.pow(query.position_delta(0), 2.0) + math.pow(query.position_delta(2), 2.0)), 0.0, 0.1) 1`] = `
[
  "surround: ",
  "math [keyword]",
  ". [access]",
  "clamp [identifier]",
  [
    "surround: ()",
    "math [keyword]",
    ". [access]",
    "sqrt [identifier]",
    [
      "surround: ()",
      "math [keyword]",
      ". [access]",
      "pow [identifier]",
      [
        "surround: ()",
        "query [keyword]",
        ". [access]",
        "position_delta [identifier]",
        [
          "surround: ()",
          "0 [number]",
        ],
        [
          "surround: ",
          ") [unknown]",
          ", [punction]",
          "2.0 [number]",
        ],
      ],
      ") [unknown]",
      "+ [operator]",
      "math [text]",
      ". [punction]",
      "pow [text]",
      [
        "surround: ()",
        "query [keyword]",
        ". [access]",
        "position_delta [identifier]",
        [
          "surround: ()",
          "2 [number]",
        ],
        [
          "surround: ",
          ") [unknown]",
          ", [punction]",
          "2.0 [number]",
        ],
      ],
      [
        "surround: ",
        ") [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      ", [punction]",
      "0.0 [number]",
      ", [punction]",
      "0.1 [number]",
    ],
  ],
  [
    "surround: ",
    ") [unknown]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 9. -175 * math.sin(variable.attack_time * 180) 1`] = `
[
  "surround: ",
  "- [operator]",
  "175 [number]",
  "* [operator]",
  "math [text]",
  ". [punction]",
  "sin [text]",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "attack_time [identifier]",
    "* [operator]",
    "180 [number]",
  ],
  [
    "surround: ",
    ") [unknown]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 10. variable.is_enchanted ? Material.enchanted : Material.default 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "is_enchanted [identifier]",
  ],
  "? [compare]",
  [
    "surround: ()",
    "Material [text]",
    ". [punction]",
    "enchanted [text]",
    ": [operator]",
    "Material [text]",
    ". [punction]",
    "default [text]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 11. variable.has_trim ? variable.trim_path : Texture.default 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "has_trim [identifier]",
  ],
  "? [compare]",
  [
    "surround: ()",
    "variable [text]",
    ". [punction]",
    "trim_path [unknown]",
    ": [operator]",
    "Texture [text]",
    ". [punction]",
    "default [text]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 12. !variable.has_trim 1`] = `
[
  "surround: ",
  "! [operator]",
  "variable [keyword]",
  ". [access]",
  "has_trim [identifier]",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 13. array.skins[query.property('minecraft:has_nectar') + query.is_angry * 2] 1`] = `
[
  "surround: ",
  "array [keyword]",
  ". [access]",
  "skins [identifier]",
  [
    "surround: []",
    "query [keyword]",
    ". [access]",
    "property [identifier]",
    [
      "surround: ()",
      "'minecraft:has_nectar' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "+ [operator]",
      "query [text]",
      ". [punction]",
      "is_angry [unknown]",
      "* [operator]",
      "2 [number]",
    ],
  ],
  [
    "surround: ",
    "] [unknown]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 14. variable.x_head_rotation = math.clamp(query.target_x_rotation, -32.5, 32.5); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "x_head_rotation [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "math [text]",
    ". [punction]",
    "clamp [text]",
    [
      "surround: ()",
      "query [keyword]",
      ". [access]",
      "target_x_rotation [identifier]",
      ", [punction]",
      "- [operator]",
      "32.5 [number]",
      ", [punction]",
      "32.5 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 15. variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "y_head_rotation [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "math [text]",
    ". [punction]",
    "clamp [text]",
    [
      "surround: ()",
      "query [keyword]",
      ". [access]",
      "target_y_rotation [identifier]",
      ", [punction]",
      "- [operator]",
      "22.5 [number]",
      ", [punction]",
      "25 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 16. variable.is_rolled_up = query.property('minecraft:armadillo_state') != 'unrolled'; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "is_rolled_up [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "property [text]",
    [
      "surround: ()",
      "'minecraft:armadillo_state' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "!= [compare]",
      "'unrolled' [text]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 17. variable.rolled_up_time = variable.is_rolled_up ? ((variable.rolled_up_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "rolled_up_time [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "variable [text]",
    ". [punction]",
    "is_rolled_up [unknown]",
    "? [compare]",
    [
      "surround: ()",
      [
        "surround: ()",
        "variable [keyword]",
        ". [access]",
        "rolled_up_time [identifier]",
        "?? [operator]",
        "0.0 [number]",
      ],
      [
        "surround: ",
        ") [unknown]",
        "+ [operator]",
        "query [text]",
        ". [punction]",
        "delta_time [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      ": [operator]",
      "0.0 [number]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 18. variable.is_peeking = query.property('minecraft:armadillo_state') == 'rolled_up_peeking'; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "is_peeking [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "property [text]",
    [
      "surround: ()",
      "'minecraft:armadillo_state' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "== [compare]",
      "'rolled_up_peeking' [text]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 19. variable.is_unrolling = query.property('minecraft:armadillo_state') == 'rolled_up_unrolling'; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "is_unrolling [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "property [text]",
    [
      "surround: ()",
      "'minecraft:armadillo_state' [text]",
    ],
    [
      "surround: ",
      ") [unknown]",
      "== [compare]",
      "'rolled_up_unrolling' [text]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 20. variable.unrolling_time = variable.is_unrolling ? ((variable.unrolling_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "unrolling_time [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "variable [text]",
    ". [punction]",
    "is_unrolling [unknown]",
    "? [compare]",
    [
      "surround: ()",
      [
        "surround: ()",
        "variable [keyword]",
        ". [access]",
        "unrolling_time [identifier]",
        "?? [operator]",
        "0.0 [number]",
      ],
      [
        "surround: ",
        ") [unknown]",
        "+ [operator]",
        "query [text]",
        ". [punction]",
        "delta_time [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      ": [operator]",
      "0.0 [number]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 21. variable.use_rolled_up_model = variable.rolled_up_time >= 0.2083 && (variable.unrolling_time == 0.0 || variable.unrolling_time <= 1.25 ); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "use_rolled_up_model [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "variable [text]",
    ". [punction]",
    "rolled_up_time [unknown]",
    ">= [compare]",
    "0.2083 [number]",
    "&& [operator]",
    [
      "surround: ()",
      "variable [keyword]",
      ". [access]",
      "unrolling_time [identifier]",
      "== [compare]",
      "0.0 [number]",
      "|| [operator]",
      "variable [text]",
      ". [punction]",
      "unrolling_time [unknown]",
      "<= [compare]",
      "1.25 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 22. variable.walking = query.modified_move_speed > 0.01 && !variable.is_rolled_up; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "walking [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "modified_move_speed [unknown]",
    "> [compare]",
    "0.01 [number]",
    "&& [operator]",
    "! [operator]",
    "variable [keyword]",
    ". [access]",
    "is_rolled_up [identifier]",
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 23. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time; 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "walk_anim_time_update [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "anim_time [unknown]",
    "+ [operator]",
    "math [text]",
    ". [punction]",
    "min [text]",
    [
      "surround: ()",
      "3.0 [number]",
      ", [punction]",
      "math [text]",
      ". [punction]",
      "lerp [text]",
      [
        "surround: ()",
        "2.0 [number]",
        ", [punction]",
        "5.0 [number]",
        ", [punction]",
        "query [text]",
        ". [punction]",
        "modified_move_speed [unknown]",
      ],
      [
        "surround: ",
        ") [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      "* [operator]",
      "query [text]",
      ". [punction]",
      "delta_time [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 24. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time;variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "walk_anim_time_update [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "query [text]",
    ". [punction]",
    "anim_time [unknown]",
    "+ [operator]",
    "math [text]",
    ". [punction]",
    "min [text]",
    [
      "surround: ()",
      "3.0 [number]",
      ", [punction]",
      "math [text]",
      ". [punction]",
      "lerp [text]",
      [
        "surround: ()",
        "2.0 [number]",
        ", [punction]",
        "5.0 [number]",
        ", [punction]",
        "query [text]",
        ". [punction]",
        "modified_move_speed [unknown]",
      ],
      [
        "surround: ",
        ") [unknown]",
      ],
    ],
    [
      "surround: ",
      ") [unknown]",
      "* [operator]",
      "query [text]",
      ". [punction]",
      "delta_time [unknown]",
    ],
  ],
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 24. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time;variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 2`] = `
[
  "surround: ",
  [
    "surround: ()",
    "variable [keyword]",
    ". [access]",
    "y_head_rotation [identifier]",
  ],
  "= [assignment]",
  [
    "surround: ()",
    "math [text]",
    ". [punction]",
    "clamp [text]",
    [
      "surround: ()",
      "query [keyword]",
      ". [access]",
      "target_y_rotation [identifier]",
      ", [punction]",
      "- [operator]",
      "22.5 [number]",
      ", [punction]",
      "25 [number]",
    ],
    [
      "surround: ",
      ") [unknown]",
    ],
  ],
]
`;
